swim/heal_partition.go

```diff
@@ -0,0 +1,154 @@
+// membership A gets merged with membership B. These need to reincarnate
+// first, before we can heal the partition with a merge.
+func nodesThatNeedToReincarnate(A, B []Change) (changesForA, changesForB []Change) {
+	// Find changes that are alive for B and faulty for A and visa versa.
+	for _, b := range B {
+		a, ok := selectMember(A, b.Address)
+		if !ok {
+			continue
+		}
+
+		// faulty for partition A, alive for partition B
+		// needs reincarnation in partition B.
+
+		if statePrecedence(a.Status) >= statePrecedence(Faulty) &&
+			b.Status == Alive &&
+			a.Incarnation >= b.Incarnation {
+
+			changesForB = append(changesForB, Change{
+				Address:     a.Address,
+				Incarnation: a.Incarnation,
+				Status:      Suspect,
+			})
+		}
+
+		// alive for partition A, faulty for partition B
+		// needs reincarnation in partition A.
+		if a.Status == Alive &&
+			statePrecedence(b.Status) >= statePrecedence(Faulty) &&
+			a.Incarnation <= b.Incarnation {
+
+			changesForA = append(changesForA, Change{
+				Address:     b.Address,
+				Incarnation: b.Incarnation,
+				Status:      Suspect,
+			})
+		}
+	}
+
+	return changesForA, changesForB
+}
+
+// reincarnateNodes applies changesForA to this nodes membership, and sends a ping
+// with changesForB to B's membership, so that B will apply those changes in its
+// ping handler.
+func reincarnateNodes(node *Node, target string, changesForA, changesForB []Change) error {
+	// reincarnate all nodes by disseminating that they are suspect
+	node.healer.logger.WithField("target", target).Info("reincarnate nodes before we can merge the partitions")
+	node.memberlist.Update(changesForA)
+
+	var err error
+	if len(changesForB) != 0 {
+		_, err = sendPingWithChanges(node, target, changesForB, time.Second)
+	}
+
+	return err
+}
+
+// mergePartitions applies the membership of B to a and send the membership
+// A to B piggybacked on top of a ping.
+func mergePartitions(node *Node, target string, B []Change) error {
+	node.healer.logger.WithField("target", target).Info("merge two partitions")
+
+	// Add membership of B to this node, so that the membership
+	// information of B will be disseminated through A.
+	node.memberlist.Update(B)
+
+	// Send membership of A to the target node, so that the membership
+	// information of partition A will be disseminated through B.
+	A := node.disseminator.MembershipAsChanges()
+	_, err := sendPingWithChanges(node, target, A, time.Second)
+	return err
+}
+
+// aliveHosts returns the address of those changes that have the Alive status.
+func aliveHosts(changes []Change) (ret []string) {
+	for _, b := range changes {
+		if b.Status == Alive {
+			ret = append(ret, b.Address)
+		}
+	}
+	return
+}
+
+// selectMember selects the member with the specified address from the partition.
+func selectMember(partition []Change, address string) (Change, bool) {
+	for _, m := range partition {
+		if m.Address == address {
+			return m, true
+		}
+	}
+
+	return Change{}, false
+}
```

Comment:
I really like the procedural approach for this algorithm (even if imperative). I think this makes the algorithm easy to follow and removes the philosophical questions of what object each method belongs to and how to share state.